### ISSUE TITLE
[prop-types] Props validation with TS StatelessComponent (SFC) type fails

### ISSUE THREAD

**Issue by yanneves** [yannickcr/eslint-plugin-react#2275]
A typical pattern for parent components fails with the error, `'children' is missing in props validation  react/prop-types`.

```typescript
import React, { StatelessComponent } from 'react'

const Wrapper: StatelessComponent = ({ children }) => (
  <div style={{ margin: '1rem' }}>{children}</div>
)
```

This looks okay to me. Presumably, the resulting JavaScript doesn't produce the validation `eslint-plugin-react` expects. Any ideas what's going on, and whether it's a bug or how it can be avoided?

**Comment by ljharb** [comment]
You’re missing:
```jsx
Wrapper.propTypes = {
  children: PropTypes.node,
};
```

**Comment by yanneves** [comment]
Thanks @ljharb, however, we avoid the need for declaring prop types this way using TypeScript, no?

Without TypeScript:
```jsx
import React from 'react'
import PropTypes from 'prop-types'

const Wrapper = ({ children }) => (
  <div style={{ margin: '1rem' }}>{children}</div>
)

Wrapper.propTypes = {
  children: PropTypes.node
}
```

And with TypeScript, as above. The `StatelessComponent` type enforces the same as `children: PropTypes.node` but on compile time afaik.

**Comment by ljharb** [comment]
No, you don’t. Types are at build-time only; propTypes run at runtime.

Separately, altho not in this or every case, propTypes are much more powerful and expressive than types. You should be using both TS types and propTypes.

(Separately, the type for “children” in a stateless component is “any”, not “node” - only you (and not the type) can know you’re passing the children prop as a child of a div, thus requiring it to be a node.

If you prefer to sacrifice a major benefit in exchange for your type system, then you’ll need to disable the prop-types rule, whose purpose is to force you to use propTypes.

**Comment by yanneves** [comment]
Gotcha! Well, that adds up - today I learned 💯

**Comment by LordZardeck** [comment]
I just came across this, and I'm not sure I understand your argument @ljharb. This currently passes eslint validation:

```js
import React, { PureComponent, ReactNode } from 'react';
import styles from './empty-state.scss';
import classNames from 'classnames';

type Props = {
    isLoading: boolean,
    children: ReactNode,
};

export default class EmptyState extends PureComponent<Props> {
    render() {
        return (
            <div className={classNames(styles.container, this.props.isLoading && styles.loadingState)}>
                {this.props.children}
            </div>
        );
    }
}

```

With what you said, I would have imagined classes failing as well. Are StatelessComponents (SFC) something special/different that they would be treated differently?

I'm wanting to enforce types being set, but not enforce the `PropTypes` implementation. (Mostly because I feel the extra code to write out twice just isn't worth it). Honestly, I think it would be neat if babel could transform the `type`s to `PropTypes`. Then you get the best of both worlds.

**Comment by LordZardeck** [comment]
Also, according to [React's Documentation](https://reactjs.org/docs/typechecking-with-proptypes.html), it's only checked at development time, so it seems to me using only Types works just fine as it won't matter if it's run-time or build-time. Am I misunderstanding something?

**Comment by ljharb** [comment]
Yes, classes are different than SFCs.

propTypes are a runtime concern, TS types are a build time concern. the difference always matters, even if the runtime checking is only in non-production

**Comment by LordZardeck** [comment]
@ljharb Can you explain the difference? I'm confused how it's okay for classes, but not SFC's.

**Comment by ljharb** [comment]
@LordZardeck TypeScript decides this, not me or JS - in your class, you do `<Props>` directly inside the class construct, which means that the named class is typed without needing inference. However, when you use an *arrow function* - as opposed to a normal, proper, named function - you've attached the type to the *variable* and not to the function itself, and the TS eslint parser seems not to pass the type info to us when you do that.

This is just another in a long list of reasons to *not use arrow functions for functional components*.

**Comment by rpivo** [comment]
> This is just another in a long list of reasons to not use arrow functions for functional components.
@ljharb Is there a resource somewhere that goes into this a bit more?

**Comment by ljharb** [comment]
Not that I'm aware of.

**Comment by ms-mousa** [comment]
> 
> This is just another in a long list of reasons to _not use arrow functions for functional components_.

Thanks for explaining this stuff. Highly appreciated. 
I managed to fix the error by defining the type of the props variable that goes into the arrow function. Like this:
```
export interface TaskCardProps {
  title: string;
  details: string;
  category: string;
}
const TaskCard: React.SFC<TaskCardProps> = (props: TaskCardProps) => {
  const { title, details, category } = props;

  return (
    ...
);
};
```


This seems to work alright. Based on what you said I'm not sacrificing any type validation, that comes from propTypes at run time. 

Can you also tell more reasons for not using arrow functions?

**Comment by ljharb** [comment]
They don’t have an explicit name - only some forms of them will have an implicit name, and it’s not always intuitive which ones will get a name inferred. If there’s no name, debugging is massively inhibited.

**Comment by varandrew** [comment]
I just came across this, and i use this.
```javascript
// add this to eslintrc.js
"react/prop-types": [2, { ignore: ['children'] }]
```

**Comment by Harvtronix** [comment]
I got here from Google, so on the odd chance my findings and conclusions will help someone, I'll document them here.

If you're using exclusively react-typescript with strict type checking (no js mixed in), then **in my opinion** it is safe to disable the `react/prop-types` eslint rule and instead rely solely on strong "compile-time" type checking to prevent you from doing bad things. This does forego the runtime type checking mentioned above, but if your entire application is written in typescript, then you can't compile/deploy your app with type errors in it anyway, so the loss of runtime type checking has minimal impact. 

This applies to both arrow functional components and regular functional components.

## Example

If you write a component like so:

```
const InputThing = (props) => {
    return (
        <input value={props.value} />
    )
}
```

You'll either get an info message from the typescript compiler saying:

> Parameter 'props' implicitly has an 'any' type, but a better type may be inferred from usage. ts(7044)

Or an error message saying:

> Parameter 'props' implicitly has an 'any' type.ts(7006)

If you get the first one, you should add `"noImplicitAny": true` to your tsconfig.json file, otherwise you're not fully taking advantage of what typescript has to offer.

Given that as the baseline, let's say you wanted to use the destructured props `children` and `value`. You'd write:

```
const InputThing = ({value, children}) => {
    return (
        <input value={value} />
    )
}
```

Two errors this time. One for value and one for children. Both saying that they have implicit "any" types, which aren't allowed. So now it's time to add prop types into the mix:

```
interface InputThingProps {
    value: string
}

const InputThing: FC<InputThingProps> = ({value, children}) => {
    return (
        <input value={value} />
    )
}
```

No more errors and now we've enforced specific property types and forms everywhere that `InputThing` is used. This works because the generic-typed `FC<>` takes a props interface as its type. FC (FunctionComponent) is defined as followed:

```
interface FunctionComponent<P = {}> {
        (props: PropsWithChildren<P>, context?: any): ReactElement<any, any> | null;
        propTypes?: WeakValidationMap<P>;
        contextTypes?: ValidationMap<any>;
        defaultProps?: Partial<P>;
        displayName?: string;
    }

```

So 

## What if you try to use a non-existent prop?

If I try to use a `foo` prop on InputThing like so:

```

```

**Comment by ljharb** [comment]
@Harvtronix however, this deprives you of the massive field of things propTypes can validate that TS is incapable of - "integer", for example - so the best practice remains to always have _both_ - TS types and propTypes.

**Comment by Harvtronix** [comment]
@ljharb I'm not aware of integer in the PropTypes library. I think you'd have to write a custom prop validator to accomplish that, which you could do in ts/js inside the component itself.

I'm not opposed to defining both a props interface and PropTypes, so I agree with your advice to cover all bases. I just wanted to spell out what I believe to be the objective net gain/loss between the two.

**Comment by ljharb** [comment]
You can write your own custom propType - you can't write a TS type that checks for integers.

**Comment by mx51damon** [comment]
Just define props as part of component, eg:

`export default class ErrorBoundary extends Component<{ children: React.ReactNode }>{ ... }`
